pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,278 @@
+package logql
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+type RangeVectorMapper struct {
+	splitByInterval time.Duration
+}
+
+func NewRangeVectorMapper(interval time.Duration) (RangeVectorMapper, error) {
+	if interval <= 0 {
+		return RangeVectorMapper{}, fmt.Errorf("cannot create RangeVectorMapper with <=0 splitByInterval. Received %s", interval)
+	}
+	return RangeVectorMapper{
+		splitByInterval: interval,
+	}, nil
+}
+
+func (m RangeVectorMapper) Parse(query string) (bool, Expr, error) {
+	parsed, err := ParseExpr(query)
+	if err != nil {
+		return false, nil, err
+	}
+
+	mapped, err := m.Map(parsed)
+	if err != nil {
+		return false, nil, err
+	}
+
+	return parsed.String() == mapped.String(), mapped, err
+}
+
+func (m RangeVectorMapper) Map(expr Expr) (Expr, error) {
+	// immediately clone the passed expr to avoid mutating the original
+	expr, err := Clone(expr)
+	if err != nil {
+		return nil, err
+	}
+
+	switch e := expr.(type) {
+	case *VectorAggregationExpr:
+		return m.mapVectorAggregationExpr(e)
+	case *RangeAggregationExpr:
+		return m.mapRangeAggregationExpr(e), nil
+	case *BinOpExpr:
+		lhsMapped, err := m.Map(e.SampleExpr)
+		if err != nil {
+			return nil, err
+		}
+		rhsMapped, err := m.Map(e.RHS)
+		if err != nil {
+			return nil, err
+		}
+		lhsSampleExpr, ok := lhsMapped.(SampleExpr)
+		if !ok {
+			return nil, badASTMapping(lhsMapped)
+		}
+		rhsSampleExpr, ok := rhsMapped.(SampleExpr)
+		if !ok {
+			return nil, badASTMapping(rhsMapped)
+		}
+		e.SampleExpr = lhsSampleExpr
+		e.RHS = rhsSampleExpr
+		return e, nil
+	default:
+		return nil, errors.Errorf("unexpected expr type (%T) for ASTMapper type (%T) ", expr, m)
+	}
+}
+
+// getRangeInterval returns the interval in the range vector
+// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
+func getRangeInterval(expr SampleExpr) time.Duration {
```

Comment:
What about 
```
count_over_time({app="foo"}[10m])
/
count_over_time({app="foo"}[30m])
```

Reply:
Each operator of binary expressions is evaluated separately, so when evaluating the first it returns `10m` and when evaluating the second it returns `10m`.